File: services/dsp-service/app/pipeline/vad.py
"""Energy + zero-crossing-rate voice activity detection.

Frames the continuous (already bandpassed + denoised + normalized) audio
stream at 20ms/10ms hop, tracks an adaptive noise floor from the quietest
recent frames, and applies onset/hangover debouncing so single noisy frames
can't flip the detector — a standard VAD design (see e.g. Sohn et al.,
"A Statistically Based Voice Activity Detector") without pulling in a heavy
external VAD dependency.
"""

from __future__ import annotations

from collections import deque
from dataclasses import dataclass

import numpy as np

from app.pipeline import constants as C


@dataclass(frozen=True)
class FrameDecision:
    time: float  # seconds, start of frame within the session's audio timeline
    is_speech: bool
    energy_db: float


def _zero_crossing_rate(frame: np.ndarray) -> float:
    if len(frame) < 2:
        return 0.0
    signs = np.sign(frame)
    signs[signs == 0] = 1
    crossings = np.sum(signs[1:] != signs[:-1])
    return float(crossings) / (len(frame) - 1)


class VoiceActivityDetector:
    def __init__(self, sample_rate: int) -> None:
        self.sample_rate = sample_rate
        self.frame_len = int(sample_rate * C.VAD_FRAME_MS / 1000)
        self.hop_len = int(sample_rate * C.VAD_HOP_MS / 1000)

        self._tail = np.zeros(0, dtype=np.float32)
        self._samples_consumed = 0

        history_len = max(5, int(C.VAD_NOISE_FLOOR_WINDOW_SEC * 1000 / C.VAD_HOP_MS))
        self._quiet_energy_history: deque[float] = deque(maxlen=history_len)

        self._confirmed_speech = False
        self._onset_counter = 0
        self._hangover_counter = 0

    def process(self, chunk: np.ndarray) -> list[FrameDecision]:
        if chunk.size:
            self._tail = np.concatenate([self._tail, chunk.astype(np.float32)])

        decisions: list[FrameDecision] = []

        while len(self._tail) >= self.frame_len:
            frame = self._tail[: self.frame_len]
            frame_time = self._samples_consumed / self.sample_rate
            self._tail = self._tail[self.hop_len :]
            self._samples_consumed += self.hop_len

            rms = float(np.sqrt(np.mean(frame**2)))
            energy_db = 20.0 * np.log10(rms + C.EPS)
            zcr = _zero_crossing_rate(frame)

            if len(self._quiet_energy_history) >= 5:
                noise_floor_db = float(np.percentile(self._quiet_energy_history, C.VAD_NOISE_FLOOR_PERCENTILE))
            else:
                # Bootstrap: assume the very first frames are non-speech room
                # tone until we've accumulated enough history to estimate a
                # real floor.
                noise_floor_db = energy_db - C.VAD_ENERGY_MARGIN_DB

            raw_speech = (energy_db > noise_floor_db + C.VAD_ENERGY_MARGIN_DB) and (zcr <= C.VAD_ZCR_MAX)

            if not raw_speech:
                self._quiet_energy_history.append(energy_db)

            if self._confirmed_speech:
                if raw_speech:
                    self._hangover_counter = 0
                else:
                    self._hangover_counter += 1
                    if self._hangover_counter >= C.VAD_HANGOVER_FRAMES:
                        self._confirmed_speech = False
                        self._onset_counter = 0
            else:
                if raw_speech:
                    self._onset_counter += 1
                    if self._onset_counter >= C.VAD_ONSET_FRAMES:
                        self._confirmed_speech = True
                        self._hangover_counter = 0
                else:
                    self._onset_counter = 0

            decisions.append(FrameDecision(time=frame_time, is_speech=self._confirmed_speech, energy_db=energy_db))

        return decisions

    @property
    def hop_seconds(self) -> float:
        return self.hop_len / self.sample_rate

File: apps/phone/src/dsp/vad.ts
// Energy + zero-crossing-rate voice activity detection. Ported from
// services/dsp-service/app/pipeline/vad.py.

import * as C from './constants';

export interface FrameDecision {
  time: number; // seconds, start of frame within the session's audio timeline
  isSpeech: boolean;
  energyDb: number;
}

function zeroCrossingRate(frame: Float32Array): number {
  if (frame.length < 2) return 0;
  let crossings = 0;
  let prevSign = frame[0] >= 0 ? 1 : -1;
  for (let i = 1; i < frame.length; i++) {
    const sign = frame[i] >= 0 ? 1 : -1;
    if (sign !== prevSign) crossings++;
    prevSign = sign;
  }
  return crossings / (frame.length - 1);
}

function percentile(values: number[], p: number): number {
  if (values.length === 0) return 0;
  const sorted = [...values].sort((a, b) => a - b);
  const idx = (p / 100) * (sorted.length - 1);
  const lo = Math.floor(idx);
  const hi = Math.ceil(idx);
  if (lo === hi) return sorted[lo];
  return sorted[lo] + (sorted[hi] - sorted[lo]) * (idx - lo);
}

function rms(frame: Float32Array): number {
  let sum = 0;
  for (let i = 0; i < frame.length; i++) sum += frame[i] * frame[i];
  return Math.sqrt(sum / frame.length);
}

export class VoiceActivityDetector {
  readonly sampleRate: number;
  private readonly frameLen: number;
  private readonly hopLen: number;

  private tail = new Float32Array(0);
  private samplesConsumed = 0;

  private readonly quietEnergyHistory: number[] = [];
  private readonly historyLen: number;

  private confirmedSpeech = false;
  private onsetCounter = 0;
  private hangoverCounter = 0;

  constructor(sampleRate: number) {
    this.sampleRate = sampleRate;
    this.frameLen = Math.floor((sampleRate * C.VAD_FRAME_MS) / 1000);
    this.hopLen = Math.floor((sampleRate * C.VAD_HOP_MS) / 1000);
    this.historyLen = Math.max(5, Math.floor((C.VAD_NOISE_FLOOR_WINDOW_SEC * 1000) / C.VAD_HOP_MS));
  }

  get hopSeconds(): number {
    return this.hopLen / this.sampleRate;
  }

  process(chunk: Float32Array): FrameDecision[] {
    if (chunk.length) {
      const merged = new Float32Array(this.tail.length + chunk.length);
      merged.set(this.tail);
      merged.set(chunk, this.tail.length);
      this.tail = merged;
    }

    const decisions: FrameDecision[] = [];

    while (this.tail.length >= this.frameLen) {
      const frame = this.tail.subarray(0, this.frameLen);
      const frameTime = this.samplesConsumed / this.sampleRate;
      this.tail = this.tail.subarray(this.hopLen);
      this.samplesConsumed += this.hopLen;

      const energyDb = 20 * Math.log10(rms(frame) + C.EPS);
      const zcr = zeroCrossingRate(frame);

      const noiseFloorDb =
        this.quietEnergyHistory.length >= 5
          ? percentile(this.quietEnergyHistory, C.VAD_NOISE_FLOOR_PERCENTILE)
          : energyDb - C.VAD_ENERGY_MARGIN_DB;

      const rawSpeech = energyDb > noiseFloorDb + C.VAD_ENERGY_MARGIN_DB && zcr <= C.VAD_ZCR_MAX;

      if (!rawSpeech) {
        this.quietEnergyHistory.push(energyDb);
        if (this.quietEnergyHistory.length > this.historyLen) this.quietEnergyHistory.shift();
      }

      if (this.confirmedSpeech) {
        if (rawSpeech) {
          this.hangoverCounter = 0;
        } else {
          this.hangoverCounter++;
          if (this.hangoverCounter >= C.VAD_HANGOVER_FRAMES) {
            this.confirmedSpeech = false;
            this.onsetCounter = 0;
          }
        }
      } else {
        if (rawSpeech) {
          this.onsetCounter++;
          if (this.onsetCounter >= C.VAD_ONSET_FRAMES) {
            this.confirmedSpeech = true;
            this.hangoverCounter = 0;
          }
        } else {
          this.onsetCounter = 0;
        }
      }

      decisions.push({ time: frameTime, isSpeech: this.confirmedSpeech, energyDb });
    }

    return decisions;
  }
}
